drv/sprot-api/src/lib.rs

```diff
@@ -333,232 +422,238 @@ impl From<u8> for MsgType {
+        payload_size: usize,
+    ) -> Result<VerifiedTxMsg, SprotError> {
+        if payload_size > PAYLOAD_SIZE_MAX {
+            return Err(SprotError::Oversize);
         }
+        self.write_header(msgtype, payload_size);
+        Ok(self.write_crc(payload_size))
     }
-}
 
-#[derive(
-    Copy, Clone, Serialize, Deserialize, SerializedSize, Eq, PartialEq,
-)]
-#[repr(C, packed)]
-pub struct UpdateRspHeader {
-    pub kind: UpdateRspKind,
-    pub value: u32, // unused, value, or error code depending on kind.
-}
+    fn write_header(&mut self, msgtype: MsgType, payload_size: usize) {
+        let _ = MsgHeader::new_v1(msgtype, payload_size)
+            .unwrap_lite()
+            .serialize(&mut self.buf[..])
+            .unwrap_lite();
+    }
 
-// Hubpack doesn't serialize Option or Result but that's
-// what we want.
-// Map the various Update api results to a signle struct
-// that hubpack can handle.
-// match Result<(), UpdateError> {
-//  Ok(()) => { new(None, None) },
-//  Err(err) => { new(None, err.into()) },
-// match Result<u32, UpdateError> {
-//  Ok(block_size) => { new(Some(block_size), None) },
-//  Err(err) => { new(None, err.into()) },
-impl UpdateRspHeader {
-    pub fn new(arg: Option<u32>, err: Option<u32>) -> Self {
-        match err {
-            None => match arg {
-                None => Self {
-                    kind: UpdateRspKind::Ok,
-                    value: 0,
-                },
-                Some(value) => Self {
-                    kind: UpdateRspKind::Value,
-                    value,
-                },
-            },
-            Some(error) => Self {
-                kind: UpdateRspKind::Error,
-                value: error,
-            },
-        }
+    fn write_crc(&mut self, payload_size: usize) -> VerifiedTxMsg {
+        let crc_begin = HEADER_SIZE + payload_size;
+        let msg_bytes = &self.buf[0..crc_begin];
+        let crc = CRC16.checksum(msg_bytes);
+        let end = crc_begin + CRC_SIZE;
+        let crc_buf = &mut self.buf[crc_begin..end];
+        let _ = hubpack::serialize(crc_buf, &crc).unwrap_lite();
+        VerifiedTxMsg(end)
     }
 }
 
-const CRC16: Crc<u16> = Crc::<u16>::new(&CRC_16_XMODEM);
-pub const HEADER_SIZE: usize = <MsgHeader as SerializedSize>::MAX_SIZE;
-const PAYLOAD_CMD_SIZE: usize = 64; // Allow for struct accompanying block data.
-const PAYLOAD_BLOCK_SIZE: usize = 1024; // Allow for bulk data (arbitrary).
-const PAYLOAD_SIZE_MAX: usize = PAYLOAD_CMD_SIZE + PAYLOAD_BLOCK_SIZE;
-const CRC_SIZE: usize = <u16 as SerializedSize>::MAX_SIZE;
-pub const BUF_SIZE: usize = HEADER_SIZE + PAYLOAD_SIZE_MAX + CRC_SIZE;
-pub const MIN_MSG_SIZE: usize = HEADER_SIZE + CRC_SIZE;
-// XXX ROT FIFO size should be discovered.
-pub const ROT_FIFO_SIZE: usize = 8;
+/// A type indicating that a complete message has been successfully serialized and therefore
+/// fits in the allocated buffer.
+#[derive(Clone, Copy)]
+pub struct VerifiedTxMsg(pub usize);
```

Comment:
Aha, this doesn't contain the data, just its size.

Would we be struck down by the Lifetime Gods if this _also_ borrowed the `&mut TxMsg`, to make sure no one else could mess with it?

Reply:
Most likely... yes. I don't think the current structure of the code lends itself well to that.  I'm not super concerned about that problem, and future improvements may make this clearer.